### PR TITLE
JP-3288 use DQ plane in residual_fringe code

### LINKED ISSUE #3937: Fringe step should use DQ arrays
The step relies on np.isfinite() of the science data to determine which pixels should have the correction applied, but instead should use the DQ values.

### PR DESCRIPTION
**Pull request by jemorrison**
Resolves [JP-3288](https://jira.stsci.edu/browse/JP-3288)

Closes #3937 

This PR adds a check on the DQ array in the residual fringe correction. If the pixel has a value of DO_NOT_USE it is not used in the correction. 
 

**Checklist for maintainers**
- [x] added entry in `CHANGES.rst` within the relevant release section
- [x] updated or added relevant tests
- [ ] updated relevant documentation
- [x] added relevant milestone
- [x] added relevant label(s)
- [x] ran regression tests, post a link to the Jenkins job below.
      [How to run regression tests on a PR](https://github.com/spacetelescope/jwst/wiki/Running-Regression-Tests-Against-PR-Branches)

https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1366/

- [ ] Make sure the JIRA ticket is [resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

### CODE REVIEW

**Comment by jemorrison** [comment]
NOT READY FOR MERGING. TESTING STEP

**Comment by jemorrison** [comment]
I have tested a few files adding this extra requirement of checking the DQ mask.
More pixels are not used in correction. The additional pixels that are rejected (in my test) fall on the very edge of the slices.
In one test I did 2159 additional pixels are rejected for both channels. On this test most of the pixels were on channel 2 and along the slice edges.

**Comment by jemorrison** [comment]
Started regression test

File: CHANGES.rst [hunk 1]
```diff
@@ -1,7 +1,10 @@
 1.14.1 (unreleased)
 ===================
 
-- 
+residual_fringe
+---------------
+- Using DQ plane and exclude pixels marked as DO_NOT_USE in correction. [#8381]
```
Comment (emolter) [inline CHANGES.rst:6]:
```suggestion
- Use DQ plane to exclude pixels marked as DO_NOT_USE in correction. [#8381]
```

File: jwst/residual_fringe/residual_fringe.py [hunk 2]
```diff
@@ -83,11 +83,14 @@ def do_correction(self):
             raise ErrorNoFringeFlat("The fringe flat step has not been run on file %s",
                                     self.input_model.meta.filename)
 
-        # Remove any NaN values from the data prior to processing
+        # Remove any NaN values adn bad pixels from the data prior to processing
```
Comment (emolter) [inline jwst/residual_fringe/residual_fringe.py:86]:
```suggestion
        # Remove any NaN values and flagged pixels from the data prior to processing
```

File: jwst/residual_fringe/residual_fringe.py [hunk 3]
```diff
@@ -83,11 +83,14 @@ def do_correction(self):
...
         # Set them to 0 for the residual fringe routine
         # They will be re-added at the end
         output_data = self.model.data.copy()
-        nanval_indx = np.where(~np.isfinite(output_data))
+        dq = self.model.dq.copy()
```
Comment (emolter) [inline jwst/residual_fringe/residual_fringe.py:90]: why does the DQ array need to be copied here?  it doesn't appear that you ever modify it

**Review by emolter — COMMENTED** [review]
Looks good, a few suggested small wording changes and one question about the code

File: jwst/residual_fringe/residual_fringe.py [hunk 4]
```diff
@@ -83,11 +83,13 @@ def do_correction(self):
...
                                     self.input_model.meta.filename)
 
-        # Remove any NaN values from the data prior to processing
+        # Remove any NaN values and flagged DO_NOT_USE pixels from the data prior to processing
         # Set them to 0 for the residual fringe routine
         # They will be re-added at the end
         output_data = self.model.data.copy()
-        nanval_indx = np.where(~np.isfinite(output_data))
+        DO_NOT_USE = datamodels.dqflags.pixel["DO_NOT_USE"]
+        nanval_indx = np.where(np.logical_and(np.bitwise_and(self.model.dq, DO_NOT_USE).astype(bool),
```
Comment (drlaw1558) [inline jwst/residual_fringe/residual_fringe.py:91]: I think this needs to be np.logical_or instead if the aim is to catch anything that has non-finite SCI values or DO_NOT_USE set.  In an ideal world those would be the same, but it doesn't hurt to check both.

Comment (jemorrison) [reply jwst/residual_fringe/residual_fringe.py]: Goodness good thing you saw that.  That is the point of the entire PR. 
I do not think they are the same from a regression test I just did. Looking in more detail at the results.

**Comment by jemorrison** [comment]
@drlaw1558 
Results of regression tests:
https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1358/#showFailuresLink
The residual_fringe step shows a difference of ~ 30% of the pixels
The majority of these pixels are related to pixels are are DO_NOT_USE are now being set to NANs.  

Something is a little strange about this test data. The output after the fringe step assigns nan values the pixel with DO_NOT_USE set on CHANNEL 2. In fact the majority of DO_NOT_USE pixels have a full dq of 262661 (NO_FLAT +  NON_SCIENCE + DO_NOT_USE)  But this does not happen in CHANNEL 1 because the input file does not have DO_NOT_USE set (except for bad pixel or something like that). The gaps on channel 1 are not assigned a value of DO_NOT_USE at the start of the residual fringe. David does that make sense to you  ?
Here is what the output looks like after residual_fringe correction (green are DO_NOT_USE). But worried this will confuse the users.  Suggestions. 
Screenshot 2024-03-28 at 10 30 37 AM

**Comment by drlaw1558** [comment]
@jemorrison That's a little puzzling; the flatfields (upstream of residual fringe) all NaN-out things outside of the slices in Ch1 (even though the gaps are smaller than for other channels) and set DQ=DNU at the same time.  What flatfield was used for that image above?

**Comment by jemorrison** [comment]
@drlaw1558 The regression test is using a cal file as input to the residual fringe step. The cal file was processed with version 1.2.1 of the pipeline and uses a the version of the flat 0538: Let me see if I can track down the rate file for MIRM108-SHORT-6021192005_1_495_MIRIFUSHORT and re-run calspec2 and then apply the residual fringe correction. It seems to be a very old flat

**Comment by drlaw1558** [comment]
Eek, that would be good to replace for sure.  I'd say try it with jw01523001001_03101_00001_mirifulong_cal.fits which you should be able to get from MAST.  That's a bright nebulae with lots of extended structure where the 2d residual fringe should be able to do a good job.

**Comment by jemorrison** [comment]
@drlaw1558 @emolter 
I changed the regression test to use a flight reference file. I made the truth file on my computer so the regression test
show small differences of running the results on two different computers. 
There was one failure because of the CHANGE log and I fixed that.
So I think this PR is ready.
https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1366/

File: CHANGES.rst [hunk 5]
```diff
@@ -1,11 +1,18 @@
...
 ===================
 
+<<<<<<< HEAD
+
 documentation
 -------------
-
 - Added docs for the NIRSpec MSA metadata file to the data products area of RTD.
   [#8399]
+=======
```
Comment (emolter) [inline CHANGES.rst:10]:
```suggestion
```

File: CHANGES.rst [hunk 6]
```diff
@@ -1,11 +1,18 @@
...
+residual_fringe
+---------------
+- Use DQ plane to exclude pixels marked as DO_NOT_USE in correction. [#8381]
+>>>>>>> 922e2c50c (updates after review)
```
Comment (emolter) [inline CHANGES.rst:14]:
```suggestion
```

File: CHANGES.rst [hunk 7]
```diff
@@ -20,6 +27,12 @@ ramp_fitting
   to use uint16 instead of uint8, in order to avoid potential
   overflow/wraparound problems. [#8377]
 
+  
+ residual_fringe
```
Comment (emolter) [inline CHANGES.rst:31]: this looks to be repeated

**Review by emolter — APPROVED** [review]
LGTM pending fixes to the change log entry.  Also, please mark this as "ready for review" instead of draft

**Comment by hbushouse** [comment]
@jemorrison can you take this out of draft form?

**Comment by jemorrison** [comment]
@hbushouse Draft form - gosh not sure how it is in draft form. I did not know we had draft form. I see it at the top - but how do I change it ?

**Comment by jemorrison** [comment]
@hbushouse @drlaw1558 I don't think the doc need update. Instead,  I think the place to put updates to docs is in flats or photom steps that have DQ = DO_NOT_USE.  For example - Jp-3587 #8385.
It is probably worth revisiting the flat step and photom step and make sure if the DQ flag set in those reference files is DO_NOT_USE then the science data is set to NAN.  That is what we want RIGHT ?
The docs for those steps should mention that.

**Review by hbushouse — APPROVED** [review]
Code and regtests look good.

**Pull request merged**